File: dashboard/src/server/hosts.ts
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import toml from "toml";
import type { HostConfig, HostsFile } from "./types";

export type { HostConfig };

// Paths are injected at module init to keep this module testable.
let configPath: string;
let hostsStorePath: string;

export function initHostPaths(opts: { configPath: string; hostsStorePath: string }) {
  configPath = opts.configPath;
  hostsStorePath = opts.hostsStorePath;
}

// Load hosts from JSON store, falling back to TOML config.
export async function loadHosts(): Promise<HostConfig[]> {
  try {
    const raw = await readFile(hostsStorePath, "utf8");
    if (!raw.trim()) {
      return [];
    }
    const parsed = JSON.parse(raw) as HostConfig[];
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    // Fallback to TOML if no JSON store exists yet.
    try {
      const raw = await readFile(configPath, "utf8");
      const data = toml.parse(raw) as HostsFile;
      const hosts = Array.isArray(data.host) ? data.host : [];
      if (hosts.length) {
        await persistHosts(hosts);
      }
      return hosts;
    } catch {
      return [];
    }
  }
}

// Find a host by ID.
export async function findHost(id: string): Promise<HostConfig | undefined> {
  const hosts = await loadHosts();
  return hosts.find((host) => host.id === id);
}

// Persist the host list to the JSON store.
export async function persistHosts(hosts: HostConfig[]) {
  const dir = path.dirname(hostsStorePath);
  await mkdir(dir, { recursive: true });
  await writeFile(hostsStorePath, JSON.stringify(hosts, null, 2));
}

// Validate a host payload, returning an ok/error discriminated union.
export function validateHost(payload: Partial<HostConfig>): { ok: true } | { ok: false; error: string } {
  if (!payload.id?.trim()) {
    return { ok: false, error: "id is required" };
  }
  if (!payload.name?.trim()) {
    return { ok: false, error: "name is required" };
  }
  if (!payload.base_url?.trim()) {
    return { ok: false, error: "base_url is required" };
  }
  return { ok: true };
}

File: crates/aiman_dashboard/src/hosts.rs
//! Host configuration loading and persistence.

use std::path::Path;

use crate::error::DashboardError;
use crate::types::{HostConfig, HostsFile};

/// Load hosts from JSON store, falling back to TOML config.
///
/// # Errors
///
/// Returns an error if neither file can be read or parsed.
pub async fn load_hosts(
    json_path: &Path,
    toml_path: &Path,
) -> Result<Vec<HostConfig>, DashboardError> {
    // Try JSON store first
    if let Ok(content) = tokio::fs::read_to_string(json_path).await {
        if !content.trim().is_empty() {
            if let Ok(hosts) = serde_json::from_str::<Vec<HostConfig>>(&content) {
                return Ok(hosts);
            }
        }
    }

    // Fallback to TOML config
    if let Ok(content) = tokio::fs::read_to_string(toml_path).await {
        let data: HostsFile = toml::from_str(&content).map_err(|e| {
            DashboardError::ConfigError(format!("invalid TOML: {e}"))
        })?;
        let hosts = data.host.unwrap_or_default();
        if !hosts.is_empty() {
            // Persist to JSON store for future loads
            let _ = persist_hosts(json_path, &hosts).await;
        }
        return Ok(hosts);
    }

    Ok(Vec::new())
}

/// Find a host by ID.
pub fn find_host<'a>(hosts: &'a [HostConfig], id: &str) -> Option<&'a HostConfig> {
    hosts.iter().find(|h| h.id == id)
}

/// Persist hosts to the JSON store.
///
/// # Errors
///
/// Returns an error if the file cannot be written.
pub async fn persist_hosts(path: &Path, hosts: &[HostConfig]) -> Result<(), DashboardError> {
    if let Some(dir) = path.parent() {
        tokio::fs::create_dir_all(dir).await.map_err(|e| {
            DashboardError::IoError(format!("failed to create directory: {e}"))
        })?;
    }
    let content = serde_json::to_string_pretty(hosts).map_err(|e| {
        DashboardError::Internal(format!("failed to serialize hosts: {e}"))
    })?;
    tokio::fs::write(path, content).await.map_err(|e| {
        DashboardError::IoError(format!("failed to write hosts: {e}"))
    })?;
    Ok(())
}

/// Validate a host configuration.
pub fn validate_host(host: &HostConfig) -> Result<(), DashboardError> {
    if host.id.trim().is_empty() {
        return Err(DashboardError::Validation("id is required".to_string()));
    }
    if host.name.trim().is_empty() {
        return Err(DashboardError::Validation("name is required".to_string()));
    }
    if host.base_url.trim().is_empty() {
        return Err(DashboardError::Validation("base_url is required".to_string()));
    }
    Ok(())
}
